File: Nebokrai/server/src/nets/networld/message.cpp
#include "message.h"

#include "../../public/crc32static.h"

#include <algorithm>
#include <array>
#include <bit>
#include <limits>
#include <mutex>

namespace WorldNet
{
namespace
{
constexpr std::size_t kEnvelopeSize = 12;
constexpr std::size_t kSmallRleInputLimit = 0x20'001;
constexpr std::size_t kSmallRleOutputCapacity = 0x10'0000;
constexpr std::uint32_t kMessageFamilyMask = 0xFFFF'FF00U;
std::mutex g_SendSerializer;

void AppendU32(std::vector<std::uint8_t>& output, std::uint32_t value)
{
    output.push_back(static_cast<std::uint8_t>(value));
    output.push_back(static_cast<std::uint8_t>(value >> 8U));
    output.push_back(static_cast<std::uint8_t>(value >> 16U));
    output.push_back(static_cast<std::uint8_t>(value >> 24U));
}
}

CMessage::CMessage(std::int32_t messageType)
{
    SetType(std::bit_cast<std::uint32_t>(messageType));
}

CMessage::CMessage(std::span<const std::uint8_t, kHeaderSize> header,
                   std::span<const std::uint8_t> payload,
                   std::uint32_t receiveTimeMs)
    : CBaseMessage(header, payload), m_ReceiveTimeMs(receiveTimeMs)
{
}

CreateMessageResult CMessage::Create(std::span<const std::uint8_t> compressed,
                                     std::uint32_t receiveTimeMs)
{
    if (compressed.empty()) {
        return {nullptr, CreateMessageError{CreateMessageErrorKind::EmptyInput}};
    }
    if (compressed.size() > std::numeric_limits<std::uint32_t>::max()) {
        return {nullptr, CreateMessageError{CreateMessageErrorKind::InputOutsideLegacyRange}};
    }

    std::size_t capacity = kSmallRleOutputCapacity;
    if (compressed.size() >= kSmallRleInputLimit) {
        if (compressed.size() >
            static_cast<std::size_t>(std::numeric_limits<std::uint32_t>::max()) / 8U) {
            return {nullptr,
                    CreateMessageError{
                        CreateMessageErrorKind::RleCapacityOverflowReactionUnknown}};
        }
        capacity = compressed.size() * 8U;
    }

    RleDecodeResult decoded = CBaseMessage::DecodeRLE(compressed, capacity);
    if (!decoded) {
        return {nullptr,
                CreateMessageError{CreateMessageErrorKind::RleDecode, decoded.error}};
    }
    return CreateMessageWithoutRLE(decoded.bytes, receiveTimeMs);
}

CreateMessageResult CMessage::CreateMessageWithoutRLE(
    std::span<const std::uint8_t> wire,
    std::uint32_t receiveTimeMs)
{
    if (wire.empty()) {
        return {nullptr, CreateMessageError{CreateMessageErrorKind::EmptyInput}};
    }
    if (wire.size() > std::numeric_limits<std::uint32_t>::max()) {
        return {nullptr, CreateMessageError{CreateMessageErrorKind::InputOutsideLegacyRange}};
    }
    if (wire.size() < kHeaderSize) {
        return {nullptr,
                CreateMessageError{CreateMessageErrorKind::HeaderTooShortReactionUnknown}};
    }

    std::array<std::uint8_t, kHeaderSize> header{};
    std::copy_n(wire.begin(), kHeaderSize, header.begin());
    return {std::unique_ptr<CMessage>(new CMessage(
                std::span<const std::uint8_t, kHeaderSize>(header),
                wire.subspan(kHeaderSize),
                receiveTimeMs)),
            std::nullopt};
}

std::int32_t CMessage::MessageType() const noexcept
{
    return std::bit_cast<std::int32_t>(GetType());
}

CBaseMessage& CMessage::Base() noexcept { return *this; }

std::span<const std::uint8_t> CMessage::WireBytes() const noexcept
{
    return {Data(), DataSize()};
}

void CMessage::ApplyClientContext(const ServerClientMessageContext& context) noexcept
{
    m_SocketID = context.socketId;
    m_MapID = context.mapId;
    m_IP = context.peerIPv4;
}

void CMessage::ApplyMapID(std::int32_t mapId) noexcept { m_MapID = mapId; }
std::int32_t CMessage::MapID() const noexcept { return m_MapID; }
std::int32_t CMessage::SocketID() const noexcept { return m_SocketID; }
std::uint32_t CMessage::PeerIPv4() const noexcept { return m_IP; }
std::uint32_t CMessage::ReceiveTimeMs() const noexcept { return m_ReceiveTimeMs; }

std::variant<std::vector<std::uint8_t>, SendMessageError>
CMessage::ServerEnvelope() const
{
    if (DataSize() >
        static_cast<std::size_t>(std::numeric_limits<std::int32_t>::max()) -
            kEnvelopeSize) {
        return SendMessageError::LengthOutsideLegacyRange;
    }

    const auto totalLength = static_cast<std::int32_t>(DataSize() + kEnvelopeSize);
    const std::uint32_t lengthWord = std::bit_cast<std::uint32_t>(totalLength);
    const std::array<std::uint8_t, 4> lengthBytes{
        static_cast<std::uint8_t>(lengthWord),
        static_cast<std::uint8_t>(lengthWord >> 8U),
        static_cast<std::uint8_t>(lengthWord >> 16U),
        static_cast<std::uint8_t>(lengthWord >> 24U)};

    std::vector<std::uint8_t> envelope;
    envelope.reserve(static_cast<std::size_t>(totalLength));
    envelope.insert(envelope.end(), lengthBytes.begin(), lengthBytes.end());
    AppendU32(envelope, DataCrc32(lengthBytes));
    AppendU32(envelope, DataCrc32(WireBytes()));
    envelope.insert(envelope.end(), Data(), Data() + DataSize());
    return envelope;
}

std::variant<std::int32_t, SendMessageError>
CMessage::SendToSocket(const ServerCommandHandle* sender, std::int32_t socketId) const
{
    if (sender == nullptr) return std::int32_t{0};
    std::lock_guard lock(g_SendSerializer);
    auto envelope = ServerEnvelope();
    if (const auto* error = std::get_if<SendMessageError>(&envelope)) return *error;
    return sender->SendBySocketID(socketId, std::get<std::vector<std::uint8_t>>(envelope));
}

std::variant<std::int32_t, SendMessageError>
CMessage::SendToMapID(const ServerCommandHandle* sender, std::int32_t mapId) const
{
    if (sender == nullptr) return std::int32_t{0};
    std::lock_guard lock(g_SendSerializer);
    auto envelope = ServerEnvelope();
    if (const auto* error = std::get_if<SendMessageError>(&envelope)) return *error;
    return sender->SendByMapID(mapId, std::get<std::vector<std::uint8_t>>(envelope));
}

std::variant<std::int32_t, SendMessageError>
CMessage::SendAll(const ServerCommandHandle* sender) const
{
    if (sender == nullptr) return std::int32_t{0};
    std::lock_guard lock(g_SendSerializer);
    auto envelope = ServerEnvelope();
    if (const auto* error = std::get_if<SendMessageError>(&envelope)) return *error;
    return sender->SendAll(std::get<std::vector<std::uint8_t>>(envelope));
}

std::variant<std::int32_t, SendMessageError>
CMessage::Send(ClientSendQueue* sender, bool prioritized) const
{
    if (sender == nullptr) return std::int32_t{0};
    std::lock_guard lock(g_SendSerializer);
    auto envelope = ServerEnvelope();
    if (const auto* error = std::get_if<SendMessageError>(&envelope)) return *error;
    return sender->SendToServer(std::get<std::vector<std::uint8_t>>(envelope), prioritized, 0);
}

std::int32_t CMessage::Run(IWorldMessageHandlers& handlers)
{
    switch (GetType() & kMessageFamilyMask) {
    case 0x0003'FC00U:
    case 0x0004'FC00U:
    case 0x0005'FA00U: handlers.OnServer(*this); break;
    case 0x0004'FB00U:
    case 0x0005'FB00U: handlers.OnLog(*this); break;
    case 0x0004'FD00U:
    case 0x0006'0400U: handlers.OnGma(*this); break;
    case 0x0005'FC00U: handlers.OnPlayer(*this); break;
    case 0x0005'FD00U: handlers.OnOther(*this); break;
    case 0x0005'FF00U: handlers.OnGm(*this); break;
    case 0x0006'0000U: handlers.OnTeam(*this); break;
    case 0x0006'0100U: handlers.OnOrganizingSystem(*this); break;
    case 0x0006'0200U:
        if (handlers.WriteLogEnabled()) handlers.OnWriteLog(*this);
        break;
    case 0x0006'0300U:
    case 0x0007'FF00U: handlers.OnCountry(*this); break;
    case 0x0006'0800U: handlers.OnServerAuction(*this); break;
    case 0x0006'0900U: handlers.OnJjcSystem(*this); break;
    case 0x0015'EB00U: handlers.OnMiscAuction(*this); break;
    default: break;
    }
    return 1;
}
}

File: Nebokrai/server/rust/src/nets/networld/message.rs
//! Сообщение Login/GameServer ↔ WorldServer из `nets/networld/message.cpp`.
//! Источник контракта — точная пара WorldServer EXE/PDB.
//!
//! Create-пути сохраняют 16-byte header, нормализацию payload length и RLE
//! threshold/capacity. Все send-направления строят общий envelope
//! `[total_len, crc(total_len), crc(message), message]`; mutex сериализует
//! build и downstream send между направлениями. `Run` выбирает исходные
//! message families и всегда возвращает `1`, включая неизвестный opcode.
//! Owned buffers, `CBaseMessage` и готовые client/server queues заменяют
//! общий scratch и ручной lifetime без изменения wire или порядка отправки.

use parking_lot::Mutex;

use crate::nets::basemessage::{CBaseMessage, RleDecodeError, decode_rle};
use crate::nets::clients::ClientSendQueue;
use crate::nets::serverclient::ServerClientMessageContext;
use crate::nets::servers::ServerCommandHandle;
use crate::public::crc32static::data_crc32;

const MESSAGE_HEADER_LEN: usize = 16;
const SERVER_ENVELOPE_LEN: usize = 12;
const SMALL_RLE_INPUT_LIMIT: usize = 0x20_001;
const SMALL_RLE_OUTPUT_CAPACITY: usize = 0x10_0000;
const MESSAGE_FAMILY_MASK: u32 = 0xFFFF_FF00;

static SEND_SERIALIZER: Mutex<()> = Mutex::new(());

/// Ошибка восстановления World-сообщения из wire-буфера.
#[derive(Debug, Eq, PartialEq)]
pub(crate) enum CreateMessageError {
    /// Нулевой decode-result либо нулевой несжатый вход не создаёт сообщение.
    EmptyInput,
    /// Внутренний wire-буфер короче обязательного header.
    HeaderTooShort { actual: usize },
    /// Размер не представим 32-битным `unsigned long` исходного API.
    InputOutsideLegacyRange,
    /// Требуемая RLE-capacity не представима в 32-битном исходном диапазоне.
    RleCapacityOutsideLegacyRange,
    /// Декодер отклонил поток либо достиг malformed-границы своего владельца.
    Rle(RleDecodeError),
}

/// Ошибка построения внешнего WorldServer envelope.
#[derive(Debug, Eq, PartialEq)]
pub(crate) enum SendMessageError {
    /// Итоговая длина не представима положительным Windows `long`.
    LengthOutsideLegacyRange,
}

/// Тринадцать доказанных свободных handler-владельцев WorldServer.
pub(crate) trait WorldMessageHandlers {
    /// Возвращает достигнутый `CGame::m_Setup.bUseLogSys` для `0x60200`.
    fn write_log_enabled(&self) -> bool;

    /// Обрабатывает server-семейства `0x3FC00`, `0x4FC00` и `0x5FA00`.
    fn on_server(&mut self, message: &mut CMessage);
    /// Обрабатывает log-семейства `0x4FB00` и `0x5FB00`.
    fn on_log(&mut self, message: &mut CMessage);
    /// Обрабатывает GMA-семейства `0x4FD00` и `0x60400`.
    fn on_gma(&mut self, message: &mut CMessage);
    /// Обрабатывает player-семейство `0x5FC00`.
    fn on_player(&mut self, message: &mut CMessage);
    /// Обрабатывает other-семейство `0x5FD00`.
    fn on_other(&mut self, message: &mut CMessage);
    /// Обрабатывает GM-семейство `0x5FF00`.
    fn on_gm(&mut self, message: &mut CMessage);
    /// Обрабатывает team-семейство `0x60000`.
    fn on_team(&mut self, message: &mut CMessage);
    /// Обрабатывает organizing-system семейство `0x60100`.
    fn on_orgasys(&mut self, message: &mut CMessage);
    /// Обрабатывает разрешённое setup-ом write-log семейство `0x60200`.
    fn on_write_log(&mut self, message: &mut CMessage);
    /// Обрабатывает country-семейства `0x60300` и `0x7FF00`.
    fn on_country(&mut self, message: &mut CMessage);
    /// Обрабатывает GameServer-to-World auction семейство `0x60800`.
    fn on_server_auction(&mut self, message: &mut CMessage);
    /// Обрабатывает JJC-system семейство `0x60900`.
    fn on_jjc_system(&mut self, message: &mut CMessage);
    /// Обрабатывает MiscServer-to-World auction семейство `0x15EB00`.
    fn on_misc_auction(&mut self, message: &mut CMessage);
}

/// Владеющее World-сообщение с runtime-метаданными соединения-источника.
pub(crate) struct CMessage {
    base: CBaseMessage,
    map_id: i32,
    socket_id: i32,
    ip: u32,
    recv_time_ms: u32,
}

impl CMessage {
    /// Создаёт исходящее сообщение указанного полного типа.
    pub(crate) fn new(message_type: i32) -> Self {
        let mut header = [0; MESSAGE_HEADER_LEN];
        header[4..8].copy_from_slice(&message_type.to_le_bytes());
        Self::from_parts(header, &[], 0)
    }

    /// Декодирует legacy RLE и создаёт внутреннее World-сообщение.
    pub(crate) fn create(compressed: &[u8], recv_time_ms: u32) -> Result<Self, CreateMessageError> {
        if compressed.is_empty() {
            return Err(CreateMessageError::EmptyInput);
        }
        if compressed.len() > u32::MAX as usize {
            return Err(CreateMessageError::InputOutsideLegacyRange);
        }

        let output_capacity = if compressed.len() < SMALL_RLE_INPUT_LIMIT {
            SMALL_RLE_OUTPUT_CAPACITY
        } else {
            compressed
                .len()
                .checked_mul(8)
                .filter(|capacity| *capacity <= u32::MAX as usize)
                .ok_or(CreateMessageError::RleCapacityOutsideLegacyRange)?
        };
        let decoded = decode_rle(compressed, output_capacity).map_err(CreateMessageError::Rle)?;
        Self::create_without_rle(&decoded, recv_time_ms)
    }

    /// Создаёт входящее сообщение из несжатых header + payload.
    pub(crate) fn create_without_rle(
        wire: &[u8],
        recv_time_ms: u32,
    ) -> Result<Self, CreateMessageError> {
        if wire.is_empty() {
            return Err(CreateMessageError::EmptyInput);
        }
        if wire.len() > u32::MAX as usize {
            return Err(CreateMessageError::InputOutsideLegacyRange);
        }
        if wire.len() < MESSAGE_HEADER_LEN {
            return Err(CreateMessageError::HeaderTooShort { actual: wire.len() });
        }

        let header = wire[..MESSAGE_HEADER_LEN]
            .try_into()
            .expect("длина World header уже проверена");
        Ok(Self::from_parts(
            header,
            &wire[MESSAGE_HEADER_LEN..],
            recv_time_ms,
        ))
    }

    /// Возвращает точный полный `MsgType` из слова header `+4`.
    pub(crate) fn message_type(&self) -> i32 {
        i32::from_le_bytes(
            self.base.as_wire_bytes()[4..8]
                .try_into()
                .expect("CBaseMessage всегда содержит 16-байтовый header"),
        )
    }

    /// Меняет только полный `MsgType`, сохраняя payload и runtime metadata.
    pub(crate) fn set_message_type(&mut self, message_type: i32) {
        self.base.set_message_type(message_type);
    }

    /// Предоставляет payload-владельцам доказанные `CBaseMessage::Get/Add`.
    pub(crate) fn base_mut(&mut self) -> &mut CBaseMessage {
        &mut self.base
    }

    /// Возвращает полное внутреннее сообщение без внешнего server envelope.
    pub(crate) fn as_wire_bytes(&self) -> &[u8] {
        self.base.as_wire_bytes()
    }

    /// Присваивает metadata accepted GameServer-соединения.
    pub(crate) fn apply_client_context(&mut self, context: ServerClientMessageContext<'_>) {
        self.socket_id = context.socket_id;
        self.map_id = context.map_id;
        self.ip = context.peer_ipv4;
    }

    /// Возвращает numeric map identity источника.
    pub(crate) const fn map_id(&self) -> i32 {
        self.map_id
    }

    /// Возвращает socket ID источника.
    pub(crate) const fn socket_id(&self) -> i32 {
        self.socket_id
    }

    /// Возвращает исходное 32-битное представление peer IPv4.
    pub(crate) const fn ip(&self) -> u32 {
        self.ip
    }

    /// Возвращает wrapping millisecond tick приёма.
    pub(crate) const fn recv_time_ms(&self) -> u32 {
        self.recv_time_ms
    }

    /// Строит исходный 12-байтовый CRC-envelope.
    pub(crate) fn server_envelope(&self) -> Result<Vec<u8>, SendMessageError> {
        let total_length = self
            .base
            .as_wire_bytes()
            .len()
            .checked_add(SERVER_ENVELOPE_LEN)
            .filter(|length| *length <= i32::MAX as usize)
            .ok_or(SendMessageError::LengthOutsideLegacyRange)?;
        let total_length_bytes = (total_length as i32).to_le_bytes();

        let mut envelope = Vec::with_capacity(total_length);
        envelope.extend_from_slice(&total_length_bytes);
        envelope.extend_from_slice(&data_crc32(&total_length_bytes).to_le_bytes());
        envelope.extend_from_slice(&data_crc32(self.base.as_wire_bytes()).to_le_bytes());
        envelope.extend_from_slice(self.base.as_wire_bytes());
        Ok(envelope)
    }

    /// Ставит server envelope одному GameServer по socket ID.
    pub(crate) fn send_to_socket(
        &self,
        sender: Option<&ServerCommandHandle>,
        socket_id: i32,
    ) -> Result<i32, SendMessageError> {
        let Some(sender) = sender else {
            return Ok(0);
        };
        let _guard = SEND_SERIALIZER.lock();
        let envelope = self.server_envelope()?;
        Ok(sender.send_by_socket_id(socket_id, &envelope))
    }

    /// Ставит server envelope одному GameServer по numeric map identity.
    pub(crate) fn send_to_map_id(
        &self,
        sender: Option<&ServerCommandHandle>,
        map_id: i32,
    ) -> Result<i32, SendMessageError> {
        let Some(sender) = sender else {
            return Ok(0);
        };
        let _guard = SEND_SERIALIZER.lock();
        let envelope = self.server_envelope()?;
        Ok(sender.send_by_map_id(map_id, &envelope))
    }

    /// Ставит server envelope всем подключённым GameServer.
    pub(crate) fn send_all(
        &self,
        sender: Option<&ServerCommandHandle>,
    ) -> Result<i32, SendMessageError> {
        let Some(sender) = sender else {
            return Ok(0);
        };
        let _guard = SEND_SERIALIZER.lock();
        let envelope = self.server_envelope()?;
        Ok(sender.send_all(&envelope))
    }

    /// Ставит server envelope исходящему LoginServer client.
    pub(crate) fn send(
        &self,
        sender: Option<&ClientSendQueue>,
        prioritized: bool,
    ) -> Result<i32, SendMessageError> {
        let Some(sender) = sender else {
            return Ok(0);
        };
        let _guard = SEND_SERIALIZER.lock();
        let envelope = self.server_envelope()?;
        Ok(sender.send_to_server(&envelope, prioritized, 0))
    }

    /// Выполняет точную маршрутизацию `Run`; неизвестный тип остаётся no-op.
    pub(crate) fn run(&mut self, handlers: &mut dyn WorldMessageHandlers) -> i32 {
        match self.message_type() as u32 & MESSAGE_FAMILY_MASK {
            0x0003_FC00 | 0x0004_FC00 | 0x0005_FA00 => handlers.on_server(self),
            0x0004_FB00 | 0x0005_FB00 => handlers.on_log(self),
            0x0004_FD00 | 0x0006_0400 => handlers.on_gma(self),
            0x0005_FC00 => handlers.on_player(self),
            0x0005_FD00 => handlers.on_other(self),
            0x0005_FF00 => handlers.on_gm(self),
            0x0006_0000 => handlers.on_team(self),
            0x0006_0100 => handlers.on_orgasys(self),
            0x0006_0200 if handlers.write_log_enabled() => handlers.on_write_log(self),
            0x0006_0300 | 0x0007_FF00 => handlers.on_country(self),
            0x0006_0800 => handlers.on_server_auction(self),
            0x0006_0900 => handlers.on_jjc_system(self),
            0x0015_EB00 => handlers.on_misc_auction(self),
            _ => {}
        }
        1
    }

    fn from_parts(header: [u8; MESSAGE_HEADER_LEN], payload: &[u8], recv_time_ms: u32) -> Self {
        Self {
            base: CBaseMessage::from_header_and_payload(header, payload),
            map_id: 0,
            socket_id: 0,
            ip: 0,
            recv_time_ms,
        }
    }
}
